source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -124,6 +141,36 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
   return {action, bytes_read, end_stream};
 }
 
+void SslSocket::complete() {
+  ASSERT(async_handshake_in_progress_);
```

Comment:
I think it could be interesting to ASSERT that we're on the same thread that we were when we started the private key operation.

Reply:
I added now support for this. I had to add a function to `Dispatcher` public API, because it was bit tricky to get access to the thread id from `ssl_socket.cc`. What do you think: is this a too big change for the gain?